File: src/main/java/data/stack/RemoveDuplicateLetters_316.java
package data.stack;

import java.util.HashMap;
import java.util.Stack;

public class RemoveDuplicateLetters_316 {
  public static String removeDuplicateLetters(String s) {
    var stack = new Stack<Character>();
    var freq = new HashMap<Character, Integer>();
    for (var c : s.toCharArray()) {
      freq.put(c, freq.getOrDefault(c, 0) + 1);
    }

    for (var c : s.toCharArray()) {
      if (stack.contains(c)) {
        freq.put(c, freq.get(c) - 1);
        continue;
      }
      while (!stack.isEmpty() && stack.peek() >= c && freq.get(stack.peek()) > 1) {
        var ch = stack.pop();
        freq.put(ch, freq.get(ch) - 1);
      }
      stack.push(c);
    }

    return stack.stream().map(String::valueOf).reduce("", String::concat);
  }

  public static void main(String[] args) {
    System.out.println(removeDuplicateLetters("bcabc"));
    System.out.println(removeDuplicateLetters("cbacdcbc"));
    System.out.println(removeDuplicateLetters("cbbbcaa"));
    System.out.println(removeDuplicateLetters("abacb"));
  }

}

File: src/main/kotlin/data/structure/stack/RemoveDuplicateLetters_316.kt
package data.structure.stack

import java.util.Stack

class KtRemoveDuplicateLetters_316 {
    fun removeDuplicateLetters(s: String): String {
        val count: MutableMap<Char, Int> = HashMap()
        s.forEach { count[it] = count.getOrDefault(it, 0) + 1 }

        val maxStack = Stack<Char>()
        s.forEach { c ->
            if (maxStack.contains(c)) {
                count[c] = count[c]!! - 1
            } else {
                // discard duplicates
                while (maxStack.isNotEmpty() && maxStack.peek() >= c && count[maxStack.peek()]!! > 1) {
                    count[maxStack.peek()] = count[maxStack.peek()]!! - 1
                    maxStack.pop()
                }
                maxStack.push(c)
            }
        }

        return maxStack.joinToString("")
    }
}

fun main() {
    val remove = KtRemoveDuplicateLetters_316()
    println(remove.removeDuplicateLetters("bcabc"))
    println(remove.removeDuplicateLetters("cbacdcbc"))
    println(remove.removeDuplicateLetters("bbcaac"))
}